### PR TITLE
Public Cache API and test

### PR DESCRIPTION
**Pull request by yschimke**
Make Cache API public, and test capturing events.

### CODE REVIEW

**Comment by yschimke** [comment]
This API doesn't address the handling of the unexplained runtime exceptions due to file corruption/race conditions/multi process etc.

But it raises a few things to consider

1) Should we have unit tests for corruption cases.
2) Should we build to provably recover or consider an API to report them? (corrupted count or onCorruption).
3) Should we provide a sample to write your own decorated filesystem that avoids the multi instance or process cases? e.g. a filesystem that writes out the process name and pid and clears up under normal uses and logs if there is a mismatch? 

3 is a half baked idea, but how do we help clients be aware that something is broken?

**Comment by swankjesse** [comment]
I'd love to figure out the definitive cause here.

If it's multiple processes overlapping? I bet there's probably some patterns here we haven't considered to detect it, and then we'd need a strategy to recover if that happens. Perhaps not cache on the 2nd instance? Yuck.

If it's multiple instances in the same process? We should probably detect this through a static variable and throw an exception?

**Comment by yschimke** [comment]
What about combining the processes and instances.  An Android gist with a FileSystem that tries to enforce uniqueness by soft-locking on "appname:instancecounter", and open and close write and delete the market file.  Report by default, but optional fail on startup during dev?

**Comment by yschimke** [comment]
@swankjesse are you for exposing this in 5.0?

**Comment by swankjesse** [comment]
Yes!

**Comment by yschimke** [comment]
OK, holding off for approval.  This change is just that minimal public API.

```
class Cache(
   directory: Path,
   maxSize: Long,
   fileSystem: FileSystem
)
```

And a useful test with example logging that should be easy to build new integrating cases off.

**Comment by yschimke** [comment]
@swankjesse ping

File: okhttp/src/test/java/okhttp3/CacheTest.java [hunk 1]
```diff
@@ -2610,6 +2613,51 @@ private RecordedRequest assertConditionallyCached(MockResponse response) throws
...
+      }
+    };
+    Path path = Path.get("/cache");
+    Cache c = new Cache(path, 100000L, loggingFileSystem);
+
+    assertThat(c.directoryPath()).isEqualTo(path);
+
+    c.size();
+
+    assertThat(events).containsExactly("metadataOrNull:/cache/journal.bkp",
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/CacheTest.java:2642]: Love it.

File: okhttp/src/test/java/okhttp3/CacheTest.java [hunk 2]
```diff
@@ -2610,6 +2613,51 @@ private RecordedRequest assertConditionallyCached(MockResponse response) throws
...
+            "metadataOrNull:/cache/journal",
+            "atomicMove:/cache/journal.tmp",
+            "atomicMove:/cache/journal",
+            "appendingSink:/cache/journal");
+
+    events.clear();
+
+    c.size();
+
+    assertThat(events).containsExactly();
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/CacheTest.java:2658]: isEmpty() ?

File: okhttp/src/test/java/okhttp3/CacheTest.java [hunk 3]
```diff
@@ -2610,6 +2613,51 @@ private RecordedRequest assertConditionallyCached(MockResponse response) throws
...
+    fileSystem.createDirectories(cache.directoryPath());
+
+    FileSystem loggingFileSystem = new ForwardingFileSystem(fileSystem) {
+      @java.lang.Override
+      public Path onPathParameter(Path path, java.lang.String functionName, java.lang.String parameterName) {
+        events.add(functionName + ":" + path);
+        return path;
+      }
+
+      @java.lang.Override
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/CacheTest.java:2629]: java.lang ??

**Review by swankjesse — APPROVED** [review]

**Pull request merged**